### PR TITLE
2.0.0-rc2

### PR DESCRIPTION
**Pull request by martinothamar**
Made changes to support 2.0.0-rc2.

With that comes some of the changes discussed in the 2.0.0 issue (#17 ) , most of it based on how the [Orleans.CosmosDB repo](https://github.com/OrleansContrib/Orleans.CosmosDB) is set up.

* New namespace standard `Orleans.Storage`
* Configuration builders, options and validator
* Changed some project names so that we can consider adding support for other kinds of providers on top of Redis (maybe this repo should be renamed to Orleans.Redis?).

### CODE REVIEW

File: Tests/Orleans.Storage.Redis.Tests/StorageTests.cs [hunk 1]
```diff
@@ -116,29 +115,62 @@ public StorageTests(ClusterFixture fixture)
...
+            var grainRef = await grain.GetReference();
+
+            var stuff = await grain.Get();
+            var scheduler = TaskScheduler.Current;
 
-            await _fixture.Database.KeyDeleteAsync(key);
+            var key = $"{grainRef.ToKeyString()}|json";
+            await _fixture.Database.HashSetAsync(key, new[] { new HashEntry("etag", "derp") });
+            
+            var exception = await grain.Set("string value", 12345, now, guid, _cluster.GrainFactory.GetGrain<IJsonTestGrain>(2222));
```
Comment (martinothamar) [inline Tests/Orleans.Storage.Redis.Tests/StorageTests.cs]: Added a test for ETag as well which I can't seem to get working. For some reason Orleans times out when calling `Set` here. When debugging I can see WriteStateAsync throwing the `ETagMismatchException` correctly, and `Set` returning it, but the call is never returned to the test and so Orleans times out after 30 seconds.

Comment (galvesribeiro) [reply Tests/Orleans.Storage.Redis.Tests/StorageTests.cs]: > ` _cluster.GrainFactory.GetGrain<IJsonTestGrain>(2222)`

Shouldn't that has a `AsReference()` or something?

**Comment by martinothamar** [comment]
Tried adding `AsReference` as suggested on the failing test, but it still fails. All the other tests works fine so to me it seems like it has something to do with an exception being thrown from the storage provider

**Comment by galvesribeiro** [comment]
Ok... I cloned your branch and ran all tests with same problem happening... 

`Exception thrown: 'System.Runtime.Serialization.SerializationException' in mscorlib.dll` is your problem which happens when you are returning in the grain method an `Exception`.

I see you have a `ETagMismatchException`. This isn't necessary. Orleans has `InconsitentStateException` so you can drop it.

So, to fix your problem:

1. Drop `ETagMismatchException` and replace it with `InconsistentStateException`.
```
if (response.IsNull)
            {
                timer.Stop();
                _logger.LogError("Writing: GrainType={0} PrimaryKey={1} Grainid={2} ETag={3} to Database={4}, finished in {5} ms - Error: ETag mismatch!",
                    grainType, key, grainReference, grainState.ETag, _db.Database, timer.Elapsed.TotalMilliseconds.ToString("0.00"));
                throw new InconsistentStateException(newEtag, grainState.ETag, $"ETag mismatch - tried with ETag: {grainState.ETag}");
            }
```
2. Change `IJsonTestGrain` to just return a `Task`. You don't need to return the exception from it.
```
Task Set(string stringValue, int intValue, DateTime dateTimeValue, Guid guidValue, IJsonTestGrain grainValue);
```
3. On your test, replace the `Assert` with this:
```
await Assert.ThrowsAsync<InconsistentStateException>(() => grain.Set("string value", 12345, now, guid, otherGrain));
```

image


Also, the new .csproj which embed the Nuget metadata into the .csproj file itself. Look at the CosmosDB provider (https://github.com/OrleansContrib/Orleans.CosmosDB/blob/master/src/Orleans.Persistence.CosmosDB/Orleans.Persistence.CosmosDB.csproj) as a reference and update your project. You can drop the .nuspec file as you wont need it anymore.

I hope it help.

**Comment by martinothamar** [comment]
Thanks!

**Comment by galvesribeiro** [comment]
Great work @martinothamar! Thanks!

**Pull request merged**